### ISSUE TITLE
⏺️ Call recording 🛠️ Server API

### ISSUE THREAD

**Issue by nickvergessen** [nextcloud/spreed#8326]
See meta issue in https://github.com/nextcloud/spreed/issues/8324

- [ ] System message when the call is started and stopped
- [ ] The recording shall be uploaded to either the/a owner/moderator of the conversation or to a "recording bot" users storage and then shared back into the chat.

**Comment by marcoambrosini** [comment]
please expose `activeSince` as we need it for display the call time!

**Comment by SystemKeeper** [comment]
Can we please also define the signaling message type already? Will this be two types for started and stopped or will there be one type with a state (like raiseHand)?

**Comment by nickvergessen** [comment]
I have no clue. @danxuliu might be better to help there.

**Comment by danxuliu** [comment]
If the Nextcloud server sends the signaling message [the only fitting type would be `message`](https://github.com/strukturag/nextcloud-spreed-signaling/blob/156bc360ffeffd3969a2f96da5c6239c09a27fa9/api_backend.go#L91-L103). That is, it would be a notification similar to [the one sent when new chat messages are posted](https://github.com/nextcloud/spreed/blob/d9f3ed06b81e036250798ca54b12ba230542a72a/lib/Signaling/Listener.php#L357-L362), so for simplicity I would go with a single message type with a payload:
```
'type' => 'recording,
'recording' => [
	'enabled' => true|false,
],
```
(feel free to suggest better names :-) ).

Having said that... we might be able to send the signaling message from the recording server rather than the Nextcloud server if it acts as an external signaling server client. Unfortunately I have not been able to explore that yet (_that_ being developing an internal client for the external signaling server), so I do not know if it would be a better approach. One benefit that I can see with this approach is that the signaling message could be sent only if the recording really started, rather than when we expect it to start.

That could be solved by sending a synchronous request from the Nextcloud server to the recording server (rather than just requesting the record to start and then forget about it) and then sending the signaling message only after the successful response was received; this would delay the response to the client starting or stopping the recording, but I guess that should not be a problem (after all we use long polling for other things :-) ).

So I would go for now with the signaling message described above, although maybe it needs to be changed later (sorry if that happens :-( ). Thoughts?

**Comment by SystemKeeper** [comment]
> ```
> 'type' => 'recording,
> 'recording' => [
> 	'enabled' => true|false,
> ],
> ```
> 
> (feel free to suggest better names :-) ).

For me that works perfectly 👍 

> So I would go for now with the signaling message described above, although maybe it needs to be changed later (sorry if that happens :-( ). Thoughts?

No problem if that happens, but now I can prepare the basic stuff already 😄

**Comment by vitormattos** [comment]
Hi @SystemKeeper and @danxuliu 
About the json format, could be as the follow format to identify if is audio or video?
```json
{
    "type" : "recording_audio"|"recording_video",
    "recording" : [
	    "enabled" : true|false
    ]
}
```

Do you have another suggestion?

**Comment by danxuliu** [comment]
@vitormattos Why not to use @nickvergessen's suggestion? That is:
```
'type' => 'recording,
'recording' => [
	'enabled' => 0|1|2,
],
```
where 0 is not enabled, 1 is video and 2 is audio (like in the room attribute).

**Comment by vitormattos** [comment]
No problem. I only need a definition about what will be the format.

I suggested to split into two types to prevent use multiple values in a case of a more simple value, as boolean value, and make more readable with specific types.

**Comment by danxuliu** [comment]
Although in general I would also favour readability in this case I think that it is better to have just a single `recording` type rather than split it, as it is consistent with the value in the room data.

Besides that, having independent `recording_audio` and `recording_video` messages could look like it is possible to have a single audio recording, a single video recording, or two recordings at the same time, one for audio and another for video, while having just `recording` with an `enabled` property either to 0, 1 or 2 makes clear (once you know what those values mean :-) ) that you can have only a single recording, either with video or just audio.

Additionally, the unified message might be easier to handle from the clients.

**Comment by vitormattos** [comment]
@danxuliu  other suggestion to check if is good thinking about enabled attribute that meaning as bool and preserving this meanign.

```
'type' => 'recording,
'recording' => [
	'enabled' => bool,
	'format' => 1|2
],
```
Or, maybe, using text (audio|video) in the format.

**Comment by SystemKeeper** [comment]
If you think enabled implies a boolean, maybe we should just go with state, as with raise hands ?
Also I would vote for having one message and not multiple.

**Comment by nickvergessen** [comment]
We don't need a boolean + an integer. that just makes the client code more complicated.
Please use a single message and a single field as was outlined above already.
Let's use `state` for the name and the values are same as the PHP constants (0 = off, 1 = video, 2 = audio)
```
'type' => 'recording,
'recording' => [
	'state' => 0|1|2,
],
```

**Comment by nickvergessen** [comment]
Part 2 merged.

From the previous tasks two small things are missing:
- [ ] `callStartTime`
  - [ ] is not documented in https://github.com/nextcloud/spreed/blob/master/docs/conversation.md
  - [ ] was missed to be added to the single room request (which is used by the HPB) https://github.com/nextcloud/spreed/blob/527ad87a78c3e18e9797681c076379e0c3bbe79e/lib/Controller/RoomController.php#L452-L472
- [ ] `callRecording`
  - [ ] is not exposed in the RoomController
    - [ ] Also then needs documenting in https://github.com/nextcloud/spreed/blob/master/docs/conversation.md

**Comment by nickvergessen** [comment]
- [ ] Recordings are not automatically stopped when the call is ended

**Comment by nickvergessen** [comment]
Shall we put the last open Todo (special media tab) into s new ticket and close this main ticket for the release?

**Comment by vitormattos** [comment]
Is a good idea to create a new ticket.
The identification of what is the type attached file isn't too necessary.
Maybe if the HPB set  the record files with a very specific name is a good temporary work around.
At the backend the change is very easy but also is necessary to change all the clients.

**Comment by vitormattos** [comment]
The new ticket: https://github.com/nextcloud/spreed/issues/8663
